### PR TITLE
Feat/lw 9092 rework wallet manager

### PR DESCRIPTION
**Pull request by AngelCastilloB**
# Context

Existing WalletManager api supports only a single wallet. It has to be re-designed for multi-wallet support.
This task connects most of the SDK pieces together.

# Proposed Solution

Re-designed WalletManager api to supports multi-wallet.

### CODE REVIEW

File: packages/e2e/test/web-extension/extension/background/walletManager.ts [hunk 1]
```diff
@@ -3,22 +3,25 @@ import {
...
-  walletManagerProperties
+  walletManagerProperties,
+  walletRepositoryProperties
 } from '@cardano-sdk/web-extension';
 
-import { AsyncKeyAgent } from '@cardano-sdk/key-management';
 import { storage as WebExtensionStorage, runtime } from 'webextension-polyfill';
-import { env, logger } from '../util';
+import { Witnesser } from '@cardano-sdk/key-management';
+import { env, logger, Metadata } from "../util";
```
Comment (mkazlauskas) [inline packages/e2e/test/web-extension/extension/background/walletManager.ts:16]: Some lint errors here

File: packages/e2e/test/web-extension/extension/ui.ts [hunk 2]
```diff
@@ -198,7 +211,48 @@ const cleanupMultidelegationInfo = (multiDelegationDiv: Element) => {
...
+
+const passphraseByteArray = Uint8Array.from(
+  env.KEY_MANAGEMENT_PARAMS.passphrase.split('').map((letter) => letter.charCodeAt(0))
+);
+merge(signerManager.signDataRequest$, signerManager.transactionWitnessRequest$).subscribe((req) => {
+  logger.info('Sign request', req);
+  if (req.walletType === WalletType.InMemory) {
+    req.sign(new Uint8Array(passphraseByteArray));
+  } else {
+    req.sign();
```
Comment (mkazlauskas) [inline packages/e2e/test/web-extension/extension/ui.ts:240]: Lint error
```suggestion
    void req.sign(new Uint8Array(passphraseByteArray));
  } else {
    void req.sign();
```

File: packages/e2e/test/web-extension/extension/ui.ts [hunk 3]
```diff
@@ -198,7 +211,48 @@ const cleanupMultidelegationInfo = (multiDelegationDiv: Element) => {
...
+
+// Expose signer manager.
+exposeSignerManagerApi(
+  {
+    signerManager
+  },
+  { logger, runtime }
+);
+
+// TODO: toSerializableObj does not support serializing empty obj {}
```
Comment (mkazlauskas) [inline packages/e2e/test/web-extension/extension/ui.ts:253]:
```suggestion
```

File: packages/e2e/test/web-extension/extension/ui.ts [hunk 4]
```diff
@@ -264,6 +363,7 @@ document.querySelector(selectors.btnSignAndBuildTx)!.addEventListener('click', a
   const { body } = await builtTx.inspect();
   logger.info('Built tx', body.outputs.length);
   const { tx: signedTx } = await builtTx.sign();
+  logger.error(signedTx);
```
Comment (mkazlauskas) [inline packages/e2e/test/web-extension/extension/ui.ts:366]:
```suggestion
```

File: packages/key-management/src/types.ts [hunk 5]
```diff
@@ -222,12 +222,13 @@ export interface Witnesser {
...
    *
-   * @param txInternals The transaction body along with its hash for which the witness data is to be generated.
+   * @param transaction The transaction along with its hash for which the witness data is to be generated.
+   * @param context The witness sign transaction context
    * @param options Optional additional parameters that may influence how the witness data is generated.
    * @returns A promise that resolves to the generated witness data for the transaction.
    */
   witness(
-    txInternals: Cardano.TxBodyWithHash,
+    transaction: Serialization.Transaction,
```
Comment (mkazlauskas) [inline packages/key-management/src/types.ts:231]: Would've been nice to do this refactor in a separate commit, please see if you can do this without too much effort. Otherwise feel free to resolve this change request

File: packages/web-extension/src/walletManager/walletManagerUi.ts [hunk 6]
```diff
@@ -15,11 +14,9 @@ import { observableWalletProperties } from '../observableWallet';
  * requires a unique `walletId`, which is calculated by {@link WalletManagerUi} before passing the message to
  * {@link WalletManagerWorker}. It is not meant to be calculated or maintained by the user.
  */
-export class WalletManagerUi implements Shutdown {
+export class WalletManagerUi<Metadata extends {}> implements Shutdown {
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManagerUi.ts:17]: Let's delete this as you suggested in DM. Then we can rename WalletManagerWorker to just WalletManager

File: packages/web-extension/src/walletManager/walletManagerWorker.ts [hunk 7]
```diff
@@ -11,48 +21,85 @@ import {
...
+      )
+    );
+
     this.#walletFactory = walletFactory;
     this.#managerStorageKey = `${walletName}-activate`;
     this.#managerStorage = managerStorage;
     this.#storesFactory = storesFactory;
     this.#logger = logger;
     this.#runtime = runtime;
+    this.#signerManagerApi = consumeSignerManagerApi({ logger: this.#logger, runtime: this.#runtime });
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManagerWorker.ts:101]: For testability and separation of concerns / boundary, it's better to just take `{signerManager: SignerManagerSignApi}` as a dependency in constructor. Application is already exposing and consuming stuff, so it can do this as well when wiring things up.

File: packages/web-extension/src/walletManager/walletManagerWorker.ts [hunk 8]
```diff
@@ -11,48 +21,85 @@ import {
...
+
     this.#hostSubscription = exposeApi(
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManagerWorker.ts:103]: Same as above, application can expose it. Maybe we can just update the interface to `activeWallet$: Observable<ObservableWallet | null>;`
```suggestion
```

File: packages/web-extension/src/walletManager/walletManager.types.ts [hunk 9]
```diff
@@ -8,11 +11,15 @@ export interface WalletManagerProps {
 export interface WalletManagerActivateProps<P extends string | number = string, O = unknown> {
   /** User given name for the observableWallet being activated */
   observableWalletName: string;
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManager.types.ts:13]: This should be replaced with name in metadata, that is stored in WalletRepository. We can add a constraint to metadata,e.g. `WalletManager<Metadata extends {name: string}>`
```suggestion
```

File: packages/web-extension/src/walletManager/walletManager.types.ts [hunk 10]
```diff
@@ -8,11 +11,15 @@ export interface WalletManagerProps {
...
-   */
-  walletId: string;
+
+  /** The walletId of the wallet to activate */
+  walletId: WalletId;
+
+  accountIndex?: number;
+
+  /** The chainId of the network to activate the wallet in */
+  chainId: Cardano.ChainId;
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManager.types.ts:21]: WalletManager already knows chainId. Is it supposed to switch network when you activate? Maybe it's better to keep activation just for wallet id, and switch network separately?
```suggestion
```

File: packages/web-extension/src/walletManager/walletManager.types.ts [hunk 11]
```diff
@@ -46,11 +53,28 @@ export interface WalletManagerActivateProps<P extends string | number = string,
...
    * Reuses the store if the wallet was previously deactivated but not destroyed.
    */
   activate(props: WalletManagerActivateProps): Promise<void>;
+
+  /**
+   * Returns true if the wallet is active.
+   *
+   * @param props The properties of the wallet to check.
+   */
+  isActive(props: WalletManagerActivateProps): Promise<boolean>;
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManager.types.ts:69]: This suggestion is only valid if we keep activation functionality to set walletId only
```suggestion
  isActive(walletId: WalletId): Promise<boolean>;
```

File: packages/web-extension/src/walletManager/walletManager.types.ts [hunk 12]
```diff
@@ -46,11 +53,28 @@ export interface WalletManagerActivateProps<P extends string | number = string,
...
+
+  /**
+   * Switches the network of the active wallet.
+   *
+   * @param id The chain id of the network to switch to.
+   */
+  switchNetwork(id: Cardano.ChainId): Promise<void>;
+
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManager.types.ts:77]: It wouldn't let me add a comment for `destroy()` as there were no changes. Let's rename it to `destroyData(walletId: WalletId)`. It shouldn't deactivate, user/app can call it separately. It can throw if you try to destroy wallet that is currently active.

File: packages/web-extension/src/walletManager/walletManager.types.ts [hunk 13]
```diff
@@ -46,11 +53,28 @@ export interface WalletManagerActivateProps<P extends string | number = string,
 }
 
 export interface WalletManagerApi {
+  chainId$: Observable<Cardano.ChainId>;
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManager.types.ts:56]: I think there's currently no way to know which wallet is active
```suggestion
  chainId$: Observable<Cardano.ChainId>;
  activeWalletId$: Observable<WalletId>;
```

File: packages/web-extension/src/walletManager/walletManagerWorker.ts [hunk 14]
```diff
@@ -11,48 +21,85 @@ import {
...
+      switchMap((wallet) =>
+        wallet.genesisParameters$.pipe(
+          map(
+            (params): Cardano.ChainId => ({
+              networkId: params.networkId,
+              networkMagic: params.networkMagic
+            })
+          )
+        )
+      )
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManagerWorker.ts:92]: It would be a better/more optimal behavior, because currently it will re-emit same chain id when you activate different wallet on same network
```suggestion
      ),
      distinctUntilChanged(deepEquals)
```

File: packages/e2e/test/web-extension/extension/background/walletManager.ts [hunk 15]
```diff
@@ -28,30 +31,40 @@ export interface WalletFactoryDependencies {
...
-        stores
+        stores,
+        witnesser
       })
     ).wallet
 };
 
 const storesFactory: StoresFactory = {
-  create: ({ walletId }) => storage.createPouchDbWalletStores(walletId, { logger })
+  create: ({ name }) => storage.createPouchDbWalletStores(name, { logger })
```
Comment (mkazlauskas) [inline packages/e2e/test/web-extension/extension/background/walletManager.ts:47]: Please confirm that this is compatible with existing store name (won't re-create the stores for existing wallets)

File: packages/web-extension/test/walletManager/walletManagerWorker.test.ts [hunk 16]
```diff
@@ -41,48 +54,74 @@ const expectWalletChannelClosed = () => {
 
 describe('WalletManagerWorker', () => {
```
Comment (mkazlauskas) [inline packages/web-extension/test/walletManager/walletManagerWorker.test.ts:55]: We should probably add a couple new tests for new functionality

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: package.json [hunk 17]
```diff
@@ -88,5 +88,8 @@
     "typedoc-plugin-missing-exports": "^1.0.0",
     "typescript": "^4.7.4"
   },
-  "packageManager": "yarn@3.2.1"
+  "packageManager": "yarn@3.2.1",
+  "dependencies": {
+    "chromedriver": "^120.0.0"
```
Comment (mkazlauskas) [inline package.json:93]: Please move this dependency to e2e package

File: packages/e2e/test/web-extension/extension/background/walletManager.ts (hunk 1 above)
Comment (AngelCastilloB) [reply packages/e2e/test/web-extension/extension/background/walletManager.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/a048f61d16b6c6969bdd63ae367f853c9a269e2f

File: packages/e2e/test/web-extension/extension/ui.ts (hunk 2 above)
Comment (AngelCastilloB) [reply packages/e2e/test/web-extension/extension/ui.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/a048f61d16b6c6969bdd63ae367f853c9a269e2f

File: packages/e2e/test/web-extension/extension/ui.ts (hunk 3 above)
Comment (AngelCastilloB) [reply packages/e2e/test/web-extension/extension/ui.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/a048f61d16b6c6969bdd63ae367f853c9a269e2f

File: packages/e2e/test/web-extension/extension/ui.ts (hunk 4 above)
Comment (AngelCastilloB) [reply packages/e2e/test/web-extension/extension/ui.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/a048f61d16b6c6969bdd63ae367f853c9a269e2f

File: package.json (hunk 17 above)
Comment (AngelCastilloB) [reply package.json]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/a048f61d16b6c6969bdd63ae367f853c9a269e2f

File: packages/web-extension/src/walletManager/walletManagerUi.ts (hunk 6 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManagerUi.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/813029a3123dac04ffb0f783c6e239a3c19ae8e9

File: packages/web-extension/src/walletManager/walletManagerWorker.ts (hunk 7 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManagerWorker.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/813029a3123dac04ffb0f783c6e239a3c19ae8e9

File: packages/web-extension/src/walletManager/walletManagerWorker.ts (hunk 8 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManagerWorker.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/813029a3123dac04ffb0f783c6e239a3c19ae8e9

File: packages/web-extension/src/walletManager/walletManager.types.ts (hunk 9 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.types.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/813029a3123dac04ffb0f783c6e239a3c19ae8e9

File: packages/web-extension/src/walletManager/walletManager.types.ts (hunk 12 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.types.ts:76]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/813029a3123dac04ffb0f783c6e239a3c19ae8e9

File: packages/web-extension/src/walletManager/walletManager.types.ts (hunk 13 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.types.ts:53]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/813029a3123dac04ffb0f783c6e239a3c19ae8e9

File: packages/web-extension/src/walletManager/walletManagerWorker.ts (hunk 14 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManagerWorker.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/813029a3123dac04ffb0f783c6e239a3c19ae8e9

File: packages/e2e/test/web-extension/extension/background/walletManager.ts (hunk 15 above)
Comment (AngelCastilloB) [reply packages/e2e/test/web-extension/extension/background/walletManager.ts:51]: It should be fine, the name of the argument/props object is not used for the db name but rather its content:

```
  const baseDbName = walletName.replace(/[^\da-z]/gi, '');
  const docsDbName = `${baseDbName}Docs`;
  ````

File: packages/web-extension/src/walletManager/walletManager.types.ts (hunk 10 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.types.ts:18]: So in order to activate a wallet we need to retrieve its state from the storage and build the providers, for this we need all three arguments, walletId, chainId and account index. This was how it worked before too btw, is just that the client code would build the store id (pubkeyhash + networkId). We are now passing the pieces separated and the wallet manager is now in charge of building the storeage ID.

I also think this makes sense, the same credentials is still a different wallet if used in a different network. If we just take the wallet ID here, we would need to guess which network id and account index the client means to activate, which we could guess by looking at the currently activa wallet and reuse the same network id and account index, but we could run into problems, lets say the user has two wallets, in one wallet he has 3 accounts and in the second wallet has 1 account, what should we do here if he is currently at account 3 on the first wallet and switches to the first wallet? we could see the accounts are different and default to account 0 (or always do this), but ultimately I think Lace should be explicit about what wallet he means to activate (which credentials, which account and in which network), and also Lace knows this, because the user will trigger this action via the GUI so Lace knows exactly what wallet the user wants (credentials, network, account). I added the utility method "switchNetwork" which does this guess work and reuses the current active wallet properties and just tries to activate with a different network.

But let me know what you think, if you dont agree with this approach I can add the logic to make it work like you suggest

File: packages/web-extension/test/walletManager/walletManagerWorker.test.ts (hunk 16 above)
Comment (AngelCastilloB) [reply packages/web-extension/test/walletManager/walletManagerWorker.test.ts]: Once we decide the actual interface functionality of the wallet manager, I will then proceed to complete the unit tests as the last step

File: packages/web-extension/src/walletManager/walletManager.types.ts (hunk 13 above)
Comment (mkazlauskas) [reply packages/web-extension/src/walletManager/walletManager.types.ts:53]: Let's combine `activeWalletId$` and `activeWalletAccount$` into 1 `activeWallet$: Observable<{walletId: WalletId; accountIndex?: number}>`, because they are always set together in 1 operation. Depending on decision https://github.com/input-output-hk/cardano-js-sdk/pull/1012#discussion_r1427616805 we could also add `chainId: ChainId` to it.

File: packages/web-extension/src/walletManager/walletManager.types.ts (hunk 11 above)
Comment (mkazlauskas) [reply packages/web-extension/src/walletManager/walletManager.types.ts:68]: Maybe we should delete this method? Doesn't seem very useful since we add the observable for active wallet props

Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.types.ts:68]: agree will remove

File: packages/e2e/test/web-extension/extension/background/walletManager.ts (hunk 15 above)
Comment (mkazlauskas) [reply packages/e2e/test/web-extension/extension/background/walletManager.ts:50]: The 'walletName' here is confusing as we're using xpub (hash?). This is how it looks now: 
image

File: package.json (hunk 17 above)
Comment (mkazlauskas) [reply package.json]: It should probably be in `devDependencies` as it's only used for running the tests, and not for anything under `src/`

File: packages/web-extension/src/walletManager/walletManager.types.ts (hunk 11 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.types.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/cd24430f838afcaaa5d8271ec804d2f53a167ad4

File: packages/web-extension/src/walletManager/walletManager.types.ts (hunk 13 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.types.ts]: changes as discussed in slack, done here: https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/cd24430f838afcaaa5d8271ec804d2f53a167ad4

File: package.json (hunk 17 above)
Comment (AngelCastilloB) [reply package.json]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/bacced6cb9d380b38fb7fba77c5489f75518655e

File: packages/web-extension/test/walletManager/walletManagerWorker.test.ts (hunk 16 above)
Comment (AngelCastilloB) [reply packages/web-extension/test/walletManager/walletManagerWorker.test.ts]: added some more tests to cover the extra cases here: https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/ee775430185f2444c530835a5ce0306e00a4986c

**Review by mkazlauskas — APPROVED** [review]
Great work! :muscle: I would reword last commit when squashing to something more generic, like `feat!: rework WalletManager` and then detail it in commit body, because it's doing a lot more than just changing type of the argument

File: packages/web-extension/src/walletManager/types.ts [hunk 18]
```diff
@@ -10,7 +10,7 @@ export enum WalletType {
 }
 
 /** For BIP-32 wallets: hash of extended account public key. For script wallets: script hash */
```
Comment (mirceahasegan) [inline packages/web-extension/src/walletManager/types.ts:12]: Is the comment still relevant?

Comment (mkazlauskas) [reply packages/web-extension/src/walletManager/types.ts:12]: I think yes, but it can be a different kind of hash. If wanted to make this an OpaqueString then I think it should be a WalletId rather than hash type.

File: packages/e2e/package.json [hunk 19]
```diff
@@ -149,6 +149,7 @@
     "babel-loader": "^8.2.5",
     "blake2b-no-wasm": "2.1.4",
     "buffer": "^6.0.3",
+    "chromedriver": "^120.0.0",
```
Comment (mirceahasegan) [inline packages/e2e/package.json:152]: This dependency was removed in the past because the actual version has to be in sync with the developer environment chrome version, which can vary.
Ideally devs would have the desired chromedriver installed globally in their setup, or use an alternative solution.
If we choose to keep the version here, [this README comment](https://github.com/input-output-hk/cardano-js-sdk/blob/master/README.md#L77) should be updated to inform the need to set this version based on the developer's chrome browser version.
Sidenote: the instructions in the [README](https://github.com/input-output-hk/cardano-js-sdk/blob/master/README.md#L77) do not work 😞

Comment (mkazlauskas) [reply packages/e2e/package.json:152]: It was re-added because global installation approach no longer works. I think we should remove the instructions and try to come up with a better way to manage this in the future

Comment (mirceahasegan) [reply packages/e2e/package.json:152]: Agree. Let's replace the instructions with "Update chromedrive dependency to the local chrome version", or something in the lines of that :D

File: packages/web-extension/src/walletManager/walletManager.ts [hunk 20]
```diff
@@ -0,0 +1,388 @@
...
+    const { walletId, chainId, accountIndex } = props;
+
+    const wallets = await firstValueFrom(this.#walletRepository.wallets$);
+    const activeWallet = wallets.find((wallet) => wallet.walletId === walletId);
+
+    if (!activeWallet) {
+      throw new InvalidArgumentError('walletId', `Wallet ${walletId} not found`);
+    }
+
+    this.#deactivateWallet();
```
Comment (mirceahasegan) [inline packages/web-extension/src/walletManager/walletManager.ts:151]: There doesn't seem to be any guard against activating the same wallet twice. It would deactivate (which shuts it down), then build the witnesser and reactivate

Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.ts:151]: The methods first line has this guard which should prevent double activation:

```
    if (this.#isActive(props)) {
      return;
    }
```

File: packages/e2e/package.json (hunk 19 above)
Comment (AngelCastilloB) [reply packages/e2e/package.json:152]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/aba05830935a2d5a441802fab53ccbcd5aea4a32

File: packages/web-extension/src/walletManager/types.ts (hunk 18 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/types.ts:12]: Yes the comment explains what the contents of wallet id could be, in the end we decided to just make it a string, should we change this?

Comment (mirceahasegan) [reply packages/web-extension/src/walletManager/types.ts:12]: No need to change it. Thank you

**Review by mirceahasegan — APPROVED** [review]
Great job, @AngelCastilloB ! 🚀

File: packages/web-extension/src/walletManager/walletManager.ts [hunk 21]
```diff
@@ -0,0 +1,388 @@
...
+    // {
+    //   walletId: WalletId; <-- I.E 'da7b4795b11a79116eb5232c83d2c862', only the 16 byte length pub key hash.
+    //   accountIndex?: number; <-- optional account number
+    //   chainId: Cardano.ChainId; <-- I.E { networkId: Cardano.NetworkId.Testnet, networkMagic: 1097911063 }
+    //   provider?: { type: P; options: O };
+    // }
+    //
+    // If we find a legacy wallet id, we need to map it to the new format and save it to storage.
+    if (this.#isLegacyProps(lastActivateProps.walletId)) {
+      lastActivateProps = this.#mapLegacyProps(lastActivateProps);
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManager.ts:122]: This would make it easier to migrate, because Lace stores wallet name in UI context: it will have to create the wallet in repository from the UI.
```suggestion
      lastActivateProps = this.#mapLegacyProps(lastActivateProps);
      if (!activeWalletExistsInRepository(lastActivateProps)) return;
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/web-extension/src/walletManager/walletManager.ts (hunk 21 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.ts]: done

**Review by mkazlauskas — DISMISSED** [review]
Great work! :rocket:

**Review by mirceahasegan — DISMISSED** [review]

File: packages/web-extension/src/walletManager/walletManager.ts [hunk 22]
```diff
@@ -0,0 +1,337 @@
...
+   * Destroys all stores for the given wallet id.
+   *
+   * @param walletId The wallet id to destroy.
+   * @private
+   */
+  async #destroyWalletStores(walletId: WalletId): Promise<void> {
+    if (this.#activeWalletProps?.walletId === walletId)
+      throw new InvalidArgumentError('walletId', 'Cannot destroy active wallet');
+
+    const stores = [...this.#walletStores.keys()].filter((key) => key.includes(walletId));
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManager.ts:203]: This won't destroy the store data if it's not loaded in memory (if wallet wasn't activated since starting)

File: packages/web-extension/src/walletManager/walletManager.types.ts [hunk 23]
```diff
@@ -59,17 +73,19 @@ export interface WalletManagerApi {
...
    */
-  destroy(): Promise<void>;
+  destroyData(walletId: WalletId): Promise<void>;
 }
 
 export interface WalletFactory {
   create: (
-    props: WalletManagerActivateProps,
-    dependencies: { keyAgent: AsyncKeyAgent; stores: storage.WalletStores }
+    name: string,
```
Comment (mkazlauskas) [inline packages/web-extension/src/walletManager/walletManager.types.ts:84]: Since wallet factory is provided by Lace, it needs to know the provider and chain id. I suggest to pass through all the activate props.
```suggestion
    props: WalletManagerActivateProps & { name: string },
```

**Review by mkazlauskas — CHANGES_REQUESTED** [review]

File: packages/web-extension/src/walletManager/walletManager.ts (hunk 22 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/17d86e545285d041a6fd47c4d48cc5b1d8897c70

File: packages/web-extension/src/walletManager/walletManager.types.ts (hunk 23 above)
Comment (AngelCastilloB) [reply packages/web-extension/src/walletManager/walletManager.types.ts]: done here https://github.com/input-output-hk/cardano-js-sdk/pull/1012/commits/39951537f0ec93792820e128dfc09540d2270a1c

**Review by mkazlauskas — APPROVED** [review]
Nice work!

**Review by mirceahasegan — APPROVED** [review]
🥇
